README.md
# LibShifterBox
<a href="https://www.buymeacoffee.com/klingo" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/default-yellow.png" alt="Buy Me A Coffee"></a>

LibShifterBox, a library add-on for '[The Elder Scrolls Online](https://www.elderscrollsonline.com/ "Home - The Elder Scrolls Online")'

## Download
You can always download the latest version here: https://www.esoui.com/downloads/info2444-LibShifterBox.html


## Quick Start Example
This is a full example of how to use the LibShifterBox. \
![alt text][shifterbox-example]
\
\
In your `MyAddon.txt` file, make sure you defined LibShifterBox as a dependency:
```
## DependsOn: LibShifterBox
```
Optionally, you can also define a specific min-version:
```
## DependsOn: LibShifterBox>=000500
```
\
\
Then the initial setup of the ShifterBox needs to be done:

```lua
-- prepare the list of entries; in this case a list of item qualities in matching colour
local leftListEntries = {
    [ITEM_QUALITY_TRASH] = GetItemQualityColor(ITEM_FUNCTIONAL_QUALITY_TRASH):Colorize(GetString("SI_ITEMQUALITY", ITEM_FUNCTIONAL_QUALITY_TRASH)),
    [ITEM_QUALITY_NORMAL] = GetItemQualityColor(ITEM_FUNCTIONAL_QUALITY_NORMAL):Colorize(GetString("SI_ITEMQUALITY", ITEM_FUNCTIONAL_QUALITY_NORMAL)),
    [ITEM_QUALITY_MAGIC] = GetItemQualityColor(ITEM_FUNCTIONAL_QUALITY_MAGIC):Colorize(GetString("SI_ITEMQUALITY", ITEM_FUNCTIONAL_QUALITY_MAGIC)),
    [ITEM_QUALITY_ARCANE] = GetItemQualityColor(ITEM_FUNCTIONAL_QUALITY_ARCANE):Colorize(GetString("SI_ITEMQUALITY", ITEM_FUNCTIONAL_QUALITY_ARCANE)),
    [ITEM_QUALITY_ARTIFACT] = GetItemQualityColor(ITEM_FUNCTIONAL_QUALITY_ARTIFACT):Colorize(GetString("SI_ITEMQUALITY", ITEM_FUNCTIONAL_QUALITY_ARTIFACT)),
    [ITEM_QUALITY_LEGENDARY] = GetItemQualityColor(ITEM_FUNCTIONAL_QUALITY_LEGENDARY):Colorize(GetString("SI_ITEMQUALITY", ITEM_FUNCTIONAL_QUALITY_LEGENDARY)),
}
-- Reminder: When you use colorized texts as values, please be aware that the color-coding becomes part of the value and thus may prevent from sorting in (visualy) alphabetical order!

-- optionally, we can override the default settings
local customSettings = {
    sortBy = "key",
    leftList = {
        title = "Lefties",
        emptyListText = "None",
        rowHeight = 48,
        fontSize = 24
    },
    rightList = {
        title = "Righties",
        emptyListText = "None",
        rowHeight = 28,
        fontSize = 14
    }
}

-- create the shifterBox and anchor it to a headerControl; also we can change the dimensions
local itemQualitiesShifterBox = LibShifterBox("MyNewAddon", "ItemQualities", parentControl, customSettings)
itemQualitiesShifterBox:SetAnchor(TOPLEFT, headerControl, BOTTOMLEFT, 0, 20)
itemQualitiesShifterBox:SetDimensions(300, 200)

-- finally, the previously defined entries are added to the left list
itemQualitiesShifterBox:AddEntriesToLeftList(leftListEntries)
```
![alt text][shifterbox-example-1]

\
Additional entries can be added, or existing ones replaced:
```lua
-- this adds a new entry to the right list
itemQualitiesShifterBox:AddEntryToRightList(10, "HelloWorld")
-- this replaces the existing [Epic] entry from the left and adds it with a new value to the right list
itemQualitiesShifterBox:AddEntryToRightList(ITEM_QUALITY_ARTIFACT, "Epic-Replacement", true)
```
![alt text][shifterbox-example-2]

\
Entries can be selected by their key:
```lua
itemQualitiesShifterBox:SelectEntryByKey(ITEM_QUALITY_LEGENDARY)
itemQualitiesShifterBox:SelectEntryByKey(10)
```
![alt text][shifterbox-example-3]

\
The whole ShifterBox is set to disabled:
```lua
itemQualitiesShifterBox:SetEnabled(false)
```
![alt text][shifterbox-example-4]


***

## API Reference
  * [Create](#create)
  * [GetShifterBox](#getshifterbox)
  * [GetControl](#getcontrol)
  * [ShifterBox:GetControl](#shifterboxgetcontrol)
  * [ShifterBox:SetAnchor](#shifterboxsetanchor)
  * [ShifterBox:SetDimensions](#shifterboxsetdimensions)
  * [ShifterBox:SetEnabled](#shifterboxsetenabled)
  * [ShifterBox:SetHidden](#shifterboxsethidden)
  * [ShifterBox:ShowCategory](#shifterboxshowcategory)
  * [ShifterBox:ShowOnlyCategory](#shifterboxshowonlycategory)
  * [ShifterBox:ShowAllCategories](#shifterboxshowallcategories)
  * [ShifterBox:HideCategory](#shifterboxhidecategory)
  * [ShifterBox:SelectEntryByKey](#shifterboxselectentrybykey)
  * [ShifterBox:SelectEntriesByKey](#shifterboxselectentriesbykey)
  * [ShifterBox:UnselectAllEntries](#shifterboxunselectallentries)
  * [ShifterBox:RemoveEntryByKey](#shifterboxremoveentrybykey)
  * [ShifterBox:RemoveEntriesByKey](#shifterboxremoveentriesbykey)
  * [ShifterBox:RegisterCallback](#shifterboxregistercallback)
  * [ShifterBox:UnregisterCallback](#shifterboxunregistercallback)
  * [LeftListBox](#leftlistbox)
    * [ShifterBox:GetLeftListEntries](#shifterboxgetleftlistentries)
    * [ShifterBox:GetLeftListEntriesFull](#shifterboxgetleftlistentriesfull)
    * [ShifterBox:AddEntryToLeftList](#shifterboxaddentrytoleftlist)
    * [ShifterBox:AddEntriesToLeftList](#shifterboxaddentriestoleftlist)
    * [ShifterBox:MoveEntryToLeftList](#shifterboxmoveentrytoleftlist)
    * [ShifterBox:MoveEntriesToLeftList](#shifterboxmoveentriestoleftlist)
    * [ShifterBox:ClearLeftList](#shifterboxclearleftlist)
  * [RightListBox](#rightlistbox)
    * [ShifterBox:GetRightListEntries](#shifterboxgetrightlistentries)
    * [ShifterBox:GetRightListEntriesFull](#shifterboxgetrightlistentriesfull)
    * [ShifterBox:AddEntryToRightList](#shifterboxaddentrytorightlist)
    * [ShifterBox:AddEntriesToRightList](#shifterboxaddentriestorightlist)
    * [ShifterBox:MoveEntryToRightList](#shifterboxmoveentrytorightlist)
    * [ShifterBox:MoveEntriesToRightList](#shifterboxmoveentriestorightlist)
    * [ShifterBox:MoveAllEntriesToRightList](#shifterboxmoveallentriestorightlist)
    * [ShifterBox:ClearRightList](#shifterboxclearrightlist)

### Create
Returns a new instance of ShifterBox with the given control name. `customSettings` can optionally be provided (with individual values), otherwise the default settings are used. Furthermore the optional `anchorOptions`, `dimensionOptions`, `leftListEntries`, and `rightListEntries` can shortcut the separate calling of `shifterBox:SetAnchor`, `shifterBox:SetDimensions`, `shifterBox:AddEntriesToLeftList` and `shifterBox:AddEntriesToRightList` respectively.
```lua
local shifterBox = LibShifterBox(uniqueAddonName, uniqueShifterBoxName, parentControl, customSettings, anchorOptions, dimensionOptions, leftListEntries, rightListEntries)
```
or
```lua
local shifterBox = LibShifterBox.Create(uniqueAddonName, uniqueShifterBoxName, parentControl, customSettings, anchorOptions, dimensionOptions, leftListEntries, rightListEntries)
```
is the same as:
```lua
local shifterBox = LibShifterBox.Create(uniqueAddonName, uniqueShifterBoxName, parentControl, customSettings)
shifterBox:SetAnchor(unpack(anchorOptions))
shifterBox:SetDimensions(unpack(dimensionOptions))
shifterBox:AddEntriesToLeftList(leftListEntries)
shifterBox:AddEntriesToRightList(rightListEntries)
```

#### customSettings
Optionally custom settings can be passed on when the ShifterBox is created.
\
The following values can be set:
```lua
customSettings = {
    showMoveAllButtons = true,  -- the >> and << buttons to move all entries can be hidden if set to false
    dragDropEnabled = true,     -- entries can be moved between lsit with drag-and-drop
    sortEnabled = true,         -- sorting of the entries can be disabled
    sortBy = "value",           -- sort the list by value or key (allowed are: "value" or "key")
    leftList = {                -- list-specific settings that apply to the LEFT list
        title = "",                                         -- the title/header of the list
        rowHeight = 32,                                     -- the height of an individual row/entry
        rowTemplateName = "ShifterBoxEntryTemplate",        -- an individual XML (virtual) control can be provided for the rows/entries
        emptyListText = GetString(LIBSHIFTERBOX_EMPTY),     -- the text to be displayed if there are no entries left in the list
        fontSize = 18,                                      -- size of the font
        rowDataTypeSelectSound = SOUNDS.ABILITY_SLOTTED,    -- an optional sound to play when a row of this data type is selected
        rowOnMouseRightClick = function(rowControl, data)   -- an optional callback function when a right-click is done inside a row element (e.g. for custom context menus)
            d("LSB: OnMouseRightClick: "..tostring(data.tooltipText))   -- reading custom 'tooltipText' from 'rowSetupAdditionalDataCallback'
        end,
        rowSetupCallback = function(rowControl, data)       -- function that will be called when a control of this type becomes visible
            d("LSB: RowSetupCallback")                      -- Calls self:SetupRowEntry, then this function, finally ZO_SortFilterList.SetupRow
        end,
        rowSetupAdditionalDataCallback = function(rowControl, data) -- data can be extended with additional data during the 'rowSetupCallback'
            d("LSB: SetupAdditionalDataCallback")
            data.tooltipText = data.value
            return rowControl, data                         -- this callback function must return the rowControl and (enriched) data again
        end,
        rowResetControlCallback = function()                -- an optional callback when the datatype control gets reset
            d("LSB: RowResetControlCallback")
        end,
    },
    rightList = {               -- list-specific settings that apply to the RIGHT list
        title = "",                                         -- the title/header of the list
        rowHeight = 32,                                     -- the height of an individual row/entry
        rowTemplateName = "ShifterBoxEntryTemplate",        -- an individual XML (virtual) control can be provided for the rows/entries
        emptyListText = GetString(LIBSHIFTERBOX_EMPTY),     -- the text to be displayed if there are no entries left in the list
        fontSize = 18,                                      -- size of the font
        rowDataTypeSelectSound = SOUNDS.ABILITY_SLOTTED,    -- an optional sound to play when a row of this data type is selected
        rowOnMouseRightClick = function(rowControl, data)   -- an optional callback function when a right-click is done inside a row element (e.g. for custom context menus)
            d("LSB: OnMouseRightClick: "..tostring(data.tooltipText))   -- reading custom 'tooltipText' from 'rowSetupAdditionalDataCallback'
        end,
        rowSetupCallback = function(rowControl, data)       -- function that will be called when a control of this type becomes visible
            d("LSB: RowSetupCallback")                      -- Calls self:SetupRowEntry, then this function, finally ZO_SortFilterList.SetupRow
        end,
        rowSetupAdditionalDataCallback = function(rowControl, data) -- data can be extended with additional data during the 'rowSetupCallback'
            d("LSB: SetupAdditionalDataCallback")
            data.tooltipText = data.value
            return rowControl, data                         -- this callback function must return the rowControl and (enriched) data again
        end,
        rowResetControlCallback = function()                -- an optional callback when the datatype control gets reset
            d("LSB: RowResetControlCallback")
        end,
    },
    callbackRegister = {                                    -- directly register callback functions with any of the exposed events
        [LibShifterBox.EVENT_LEFT_LIST_ROW_ON_MOUSE_ENTER] = function(rowControl, shifterBox, data)
            d("LSB: LeftListRowOnMouseEnter")
        end,
        [LibShifterBox.EVENT_RIGHT_LIST_ROW_ON_MOUSE_ENTER] = function(rowControl, shifterBox, data)
            d("LSB: RightListRowOnMouseEnter")
        end
    }
}
```
For `rowDataTypeSelectSound`, you may look at the overview of [Sounds](https://wiki.esoui.com/Sounds) in the ESOUI Wiki.

#### anchorOptions
Optionally anchorOptions can be passed on when the ShifterBox is created. This replaces the separate call of `shifterBox:SetAnchors()`.
\
The following values can be set:
```
anchorOptions = {
    number whereOnMe,
    object anchorTargetControl,
    number whereOnTarget, 
    number offsetX, 
    number offsetY
}
```

#### dimensionOptions
Optionally dimensionOptions can be passed on when the ShifterBox is created. This replaces the separate call of `shifterBox:SetDimensions()`.
\
The following values can be set:
```
dimensionOptions = {
    number width,
    number height
}
```

#### leftListEntries
Optionally leftListEntries can be passed on when the ShifterBox is created to directly populate the left listBox. This replaces the separate call of `shifterBox:AddEntriesToLeftList()`. Note though that this way no `categoryId` can be provided for the entries.
\
The `leftListEntries` must either be a table with the following format, or a function return such table:
```
{
    [key] = value,
    [key] = value
}
```

#### rightListEntries
Optionally rightListEntries can be passed on when the ShifterBox is created to directly populate the right listBox. This replaces the separate call of `shifterBox:AddEntriesToRightList()`. Note though that this way no `categoryId` can be provided for the entries.
\
The `rightListEntries` must either be a table with the following format, or a function return such table:
```
{
    [key] = value,
    [key] = value
}
```


### GetShifterBox
Returns the (first to be created) ShifterBox instance based on the `uniqueAddonName` and `uniqueShifterBoxName`.
```lua
local shifterBox = LibShifterBox.GetShifterBox(uniqueAddonName, uniqueShifterBoxName)
```

### GetControl
Returns the CT_CONTROL object of the (first to be created) ShifterBox based on the `uniqueAddonName` and `uniqueShifterBoxName`. This can be used to e.g. anchor other controls to the ShifterBox.
Additionally, as second return parameter the instance of the shifterBox itself is returned.
\
It is preferred to use the `:GetControl()` function of your instantiated shifterBox (see below).
```lua
local shifterBoxControl, shifterBox = LibShifterBox.GetControl(uniqueAddonName, uniqueShifterBoxName)
```

### ShifterBox:GetControl
Returns the CT_CONTROL object of the instantiated shifterBox. This can be used to e.g. anchor other controls to the ShifterBox.
Additionally, as second return parameter the instance of the shifterBox itself is returned.
```lua
local shifterBoxControl, shifterBox = shifterBox:GetControl()
```

### ShifterBox:SetAnchor
Sets the anchor of the ShifterBox to any other control of your UI. Only one anchor is supported and any previous anchors will first be cleared.
```lua
shifterBox:SetAnchor(whereOnMe, anchorTargetControl, whereOnTarget, offsetX, offsetY)
```

### ShifterBox:SetDimensions
Sets the dimensions of the overall ShifterBox (across both listBoxes and including header titles if applicable). The provided width is distributed to the two listBoxes and the space in between (for the buttons). There is a minimum height/width of 80x80.
```lua
shifterBox:SetDimensions(width, height)
```

### ShifterBox:SetEnabled
Sets the whole ShifterBox to enabled or disabled state. It is no longer possible to select entries or to shift them between the two listBoxes if disabled.
```lua
shifterBox:SetEnabled(enabled)
```

### ShifterBox:SetHidden
Sets the whole ShifterBox to hidden or shows it again.
```lua
shifterBox:SetHidden(hidden)
```

### ShifterBox:ShowCategory
Shows all entries that have been added to the shifterBox under the provided `categoryId`.
```lua
shifterBox:ShowCategory(categoryId)
```

### ShifterBox:ShowOnlyCategory
Shows all entries that have been added to the shifterBox under the provided `categoryId`. All entries with a different `categoryId` will be hidden.
```lua
shifterBox:ShowOnlyCategory(categoryId)
```

### ShifterBox:ShowAllCategories
Shows all entries that have been hidden with `shifterBox:HideCategory(categoryId)`.
```lua
shifterBox:ShowAllCategories()
```

### ShifterBox:HideCategory
Hides all entries that have been added to the shifterBox under the provided `categoryId`.
```lua
shifterBox:HideCategory(categoryId)
```

### ShifterBox:SelectEntryByKey
Selects (or deselects if already selected) an entry on either listBox based on the provided key.
```lua
shifterBox:SelectEntryByKey(key)
```

### ShifterBox:SelectEntriesByKey
Selects (or deselects if already selected) a list of entries on either listBox. The provided `keys` must be a table with keys such as `keys = {1, 2, 3}`.
```lua
shifterBox:SelectEntriesByKey(keys)
```

### ShifterBox:UnselectAllEntries
Deselects all entries on either listBox.
```lua
shifterBox:UnselectAllEntries()
```

### ShifterBox:RemoveEntryByKey
Removes an entry on either listBox based on the provided key.
```lua
shifterBox:RemoveEntryByKey(key)
```

### ShifterBox:RemoveEntriesByKey
Removes a list of entries on either listBox. The provided `keys` must be a table with keys such as `keys = {1, 2, 3}`.
```lua
shifterBox:RemoveEntriesByKey(keys)
```

### ShifterBox:RegisterCallback
Register your own `callbackFunction` that is executed upon various shifterBox events.
```lua
shifterBox:RegisterCallback(shifterBoxEvent, callbackFunction)
```
The following values for `shifterBoxEvent` are currently supported:

#### LibShifterBox.EVENT_LIST_CREATED
This event is triggered when a list has been created and thus is accessible now to other functions such as for adding new entries to it.
```lua
-- @param shifterBox object referencing the shifterBox that triggered this event
-- @param isLeftList boolean indicating if it is the left list (true) or the right list (false)
-- @param listControl object referencing the list control that has been created
local function myListCreatedFunction(shifterBox, isLeftList, listControl)
    -- do something
end
shifterBox:RegisterCallback(LibShifterBox.EVENT_LIST_CREATED, myListCreatedFunction)
```

#### LibShifterBox.EVENT_LIST_CLEARED
This event is triggered when a list has been cleared from all (shown) entries, i.e. the entries have been moved to the other list, or got deleted from it. \
It does not check for entries that are part of a category that is currently hidden, only entries from shown categories are considered when evaluating if the list is cleared or not. The event however can be triggered when `ShifterBox:HideCategory` or `ShifterBox:ShowOnlyCategory` are called and the list does not have any entries left.
```lua
-- @param shifterBox object referencing the shifterBox that triggered this event
-- @param isLeftList boolean indicating if it is the left list (true) or the right list (false)
local function myListClearedFunction(shifterBox, isLeftList)
    -- do something
end
shifterBox:RegisterCallback(LibShifterBox.EVENT_LIST_CLEARED, myListClearedFunction)
```

#### LibShifterBox.EVENT_LIST_ENTRY_ADDED
This event is triggered whenever a new entry is added to the list with `ShifterBox:AddEntryToRightList` (once) or `ShifterBox:AddEntriesToRightList` (multiple times). \
It is NOT triggered when items are moved from one list to the other list.
```lua
-- @param shifterBox object referencing the shifterBox that triggered this event
-- @param isLeftList boolean indicating if it is the left list (true) or the right list (false)
-- @param list object referencing the ShifterBoxList (subclass of ZO_SortFilterList)
-- @param entryAdded table containing details about the added entry
local function myListEntryAddedFunction(shifterBox, isLeftList, list, entryAdded)
    -- do something
end
shifterBox:RegisterCallback(LibShifterBox.EVENT_LIST_ENTRY_ADDED, myListEntryAddedFunction)
```
The returned `entryAdded` has the following structure:
```lua
entryAdded = {
    key=key,
    value=value,
    categoryId=categoryId,
}
```

#### LibShifterBox.EVENT_LIST_ENTRY_REMOVED
This event is triggered whenever an existing entry is removed from the list with `ShifterBox:RemoveEntryByKey` (once) or `ShifterBox:RemoveEntriesByKey` (multiple times). It can also be triggered by `ShifterBox:AddEntryTo___List` (once) and `ShifterBox:AddEntriesTo___List` (multiple times) when `replace=true` and an entry with the same key already exists. In that case it will first be removed, and then a new entry is added. \
It is NOT triggered when items are moved from one list to the other list.
```lua
-- @param shifterBox object referencing the shifterBox that triggered this event
-- @param isLeftList boolean indicating if it is the left list (true) or the right list (false)
-- @param list object referencing the ShifterBoxList (subclass of ZO_SortFilterList)
-- @param entryRemoved table containing details about the removed entry
local function myListEntryRemovedFunction(shifterBox, isLeftList, list, entryRemoved)
    -- do something
end
shifterBox:RegisterCallback(LibShifterBox.EVENT_LIST_ENTRY_REMOVED, myListEntryRemovedFunction)
```
The returned `entryRemoved` has the following structure:
```lua
entryRemoved = {
    key=key,
}
```

#### LibShifterBox.EVENT_LIST_ENTRY_SHIFTED
This event is triggered when an entry is shifted from one list to another, either with the [<] and [>] buttons, by drag-and-drop or with any of the library functions. \
Note that when you move multiple entries, this event is also triggered multiple times (once per shifted entry).
```lua
-- @param shifterBox object referencing the shifterBox that triggered this event
-- @param isDestLeftList boolean indicating if the DESTINATION list is the left list (true) or the right list (false)
-- @param key string with the key of the moved entry
-- @param value string with the (displayed) value of the moved entry
-- @param categoryId string with the category of the moved entry (can be nil)
-- @param fromList object referencing the SOURCE ShifterBoxList (subclass of ZO_SortFilterList)
-- @param toList object referencing the DESTINATION ShifterBoxList (subclass of ZO_SortFilterList)
local function myEntryShiftedFunction(shifterBox, isDestLeftList, key, value, categoryId, fromList, toList)
    -- do something
end
shifterBox:RegisterCallback(LibShifterBox.EVENT_LIST_ENTRY_SHIFTED, myEntryShiftedFunction)
```

#### LibShifterBox.EVENT_LIST_ROW_HIGHLIGHTED

info/changelog.txt
0.5.0   (July 25, 2021)
- updated to APIVersion 100035 (Blackwood) ([URL="https://github.com/klingo/ESO-LibShifterBox/pull/24"]#24[/URL], thanks Baertram)
- added many more events and callback functions, improved drag&drop visualisation, and other code improvements ([URL="https://github.com/klingo/ESO-LibShifterBox/pull/24"]#24[/URL], thanks Baertram / [URL="https://github.com/klingo/ESO-LibShifterBox/pull/25"]#25[/URL])

0.4.1   (March 06, 2021)
- updated to APIVersion 100034 (Flames of Ambition)) ([URL="https://github.com/klingo/ESO-LibShifterBox/pull/21"]#21[/URL])
- fixed an issue where the (sorting) headers dit not get properly disabled ([URL="https://github.com/klingo/ESO-LibShifterBox/pull/22"]#22[/URL])

0.4.0   (January 24, 2021)
- added two new events: EVENT_LEFT_LIST_CLEARED, EVENT_RIGHT_LIST_CLEARED ([URL="https://github.com/klingo/ESO-LibShifterBox/pull/19"]#19[/URL])
- fixed an issue where the << and >> buttons were not enabled again after adding a new entry ([URL="https://github.com/klingo/ESO-LibShifterBox/pull/16"]#16[/URL])
- fixed an issue with moving selected entries when a new entry is added in between ([URL="https://github.com/klingo/ESO-LibShifterBox/pull/17"]#17[/URL] / [URL="https://github.com/klingo/ESO-LibShifterBox/pull/18"]#18[/URL])

0.3.4   (November 03, 2020)
- updated to APIVersion 100033 (Markarth) ([URL="https://github.com/klingo/ESO-LibShifterBox/pull/11"]#11[/URL])
- fixed an issue where non-EN localization was not applied ([URL="https://github.com/klingo/ESO-LibShifterBox/pull/10"]#10[/URL])

0.3.3   (September 05, 2020)
- updated to APIVersion 100032 (Stonethorn)

0.3.2   (April 24, 2020)
- updated to APIVersion 100031 (Graymoor)

0.3.1   (February 24, 2020)
- updated to APIVersion 100030 (Harrowstorm)

0.3.0   (December 17, 2019)
- add two new library functions: [RegisterCallback] and [UnregisterCallback]
- updated to APIVersion 100029 (Dragonhold)

0.2.0   (August 21, 2019)
- restructured customSettings (breaking change!)
- add [dragDropEnabled] setting to control the drag-and-drop feature (by Baertram)
- add [fontSize] setting to change the font of the entries
- add localization for default empty text

0.1.1   (August 18, 2019)
- add [sortEnabled] customSetting as proposed by Baertram
- add additional shortcut parameters to .Create/:New as proposed by Baertram

0.1.0   (August 18, 2019)
- initial (beta) release

ShifterBoxExample/ShifterBoxExample.lua
local _addonName = "ShifterBoxExample"

local LAM2 = LibAddonMenu2 or LibStub("LibAddonMenu-2.0")
local OptionsTable = setmetatable({}, { __index = table })
local OptionSubTable = setmetatable({}, { __index = table })
local PanelData = {
    type = "panel",
    name = "ShifterBoxExample",
    author = "Klingo",
    registerForRefresh = true,
    registerForDefaults = false,
}

local function initShifterBoxExample()
    -- prepare the list of entries; in this case a list of item qualities in matching colour
    local leftListEntries = {
        [ITEM_QUALITY_TRASH] = GetItemQualityColor(ITEM_FUNCTIONAL_QUALITY_TRASH):Colorize(GetString("SI_ITEMQUALITY", ITEM_FUNCTIONAL_QUALITY_TRASH)),
        [ITEM_QUALITY_NORMAL] = GetItemQualityColor(ITEM_FUNCTIONAL_QUALITY_NORMAL):Colorize(GetString("SI_ITEMQUALITY", ITEM_FUNCTIONAL_QUALITY_NORMAL)),
        [ITEM_QUALITY_MAGIC] = GetItemQualityColor(ITEM_FUNCTIONAL_QUALITY_MAGIC):Colorize(GetString("SI_ITEMQUALITY", ITEM_FUNCTIONAL_QUALITY_MAGIC)),
        [ITEM_QUALITY_ARCANE] = GetItemQualityColor(ITEM_FUNCTIONAL_QUALITY_ARCANE):Colorize(GetString("SI_ITEMQUALITY", ITEM_FUNCTIONAL_QUALITY_ARCANE)),
        [ITEM_QUALITY_ARTIFACT] = GetItemQualityColor(ITEM_FUNCTIONAL_QUALITY_ARTIFACT):Colorize(GetString("SI_ITEMQUALITY", ITEM_FUNCTIONAL_QUALITY_ARTIFACT)),
        [ITEM_QUALITY_LEGENDARY] = GetItemQualityColor(ITEM_FUNCTIONAL_QUALITY_LEGENDARY):Colorize(GetString("SI_ITEMQUALITY", ITEM_FUNCTIONAL_QUALITY_LEGENDARY)),
    }
    -- Reminder: When you use colorized texts as values, please be aware that the color-coding becomes part of the value and thus may prevent from sorting in (visualy) alphabetical order!

    -- optionally, we can override the default settings
    local customSettings = {
        sortBy = "key",
        leftList = {
            title = "Available",
            emptyListText = "None",
            rowHeight = 28,
            fontSize = 16
        },
        rightList = {
            title = "Selected",
            emptyListText = "None",
            rowHeight = 28,
            fontSize = 16
        }
    }

    -- create the shifterBox and anchor it to a headerControl; also we can change the dimensions
    local itemQualitiesShifterBox = LibShifterBox("MyShifterBoxExample", "ItemQualities", ShifterBoxExampleMainWindow, customSettings)
    itemQualitiesShifterBox:SetAnchor(TOPLEFT, ShifterBoxExampleMainWindowHeader, BOTTOMLEFT, 0, 20)
    itemQualitiesShifterBox:SetDimensions(300, 200)

    -- finally, the previously defined entries are added to the left list
    itemQualitiesShifterBox:AddEntriesToLeftList(leftListEntries)



    -- --------------------------------------------------------

    local clearLeftListButton = ShifterBoxExampleMainWindow:GetNamedChild("ClearLeftList")
    clearLeftListButton:SetHandler("OnClicked", function(self)
        itemQualitiesShifterBox:ClearLeftList()
    end)

    local clearRightListButton = ShifterBoxExampleMainWindow:GetNamedChild("ClearRightList")
    clearRightListButton:SetHandler("OnClicked", function(self)
        itemQualitiesShifterBox:ClearRightList()
    end)

    local addEntryToLeftButton = ShifterBoxExampleMainWindow:GetNamedChild("AddEntryToLeft")
    local addEntryToLeftListIndexCounter = 99
    addEntryToLeftButton:SetHandler("OnClicked", function(self)
        itemQualitiesShifterBox:AddEntryToLeftList(addEntryToLeftListIndexCounter, "entry_" .. tostring(addEntryToLeftListIndexCounter), true)
        addEntryToLeftListIndexCounter = addEntryToLeftListIndexCounter + 1
    end)



    -- --------------------------------------------------------

    local function myLeftListClearedFunction(shifterBox, hasStillHiddenEntries)
        df("myLEFTListClearedFunction | hasStillHiddenEntries = %s", tostring(hasStillHiddenEntries))
    end
    itemQualitiesShifterBox:RegisterCallback(LibShifterBox.EVENT_LEFT_LIST_CLEARED, myLeftListClearedFunction)

    local function myRightListClearedFunction(shifterBox, hasStillHiddenEntries)
        df("myRIGHTListClearedFunction | hasStillHiddenEntries = %s", tostring(hasStillHiddenEntries))
    end
    itemQualitiesShifterBox:RegisterCallback(LibShifterBox.EVENT_RIGHT_LIST_CLEARED, myRightListClearedFunction)

    --    local function myEntryMovedFunction(shifterBox, key, value, categoryId, isDestListLeftList)
    --        shifterBox:AddEntryToLeftList("allo", "test", true)
    --    end
    --    itemQualitiesShifterBox:RegisterCallback(LibShifterBox.EVENT_ENTRY_MOVED, myEntryMovedFunction)
end

local function initLAMMenuWithShifterBoxExample()
    local fullWidthShifterBoxDisabled = true

    OptionsTable:insert({
        type = "header",
        name = "ShifterBox Example"
    })

    OptionsTable:insert({
        type = "description",
        text = "Here it comes in full-width:"
    })

    OptionsTable:insert({
        type = "shifterbox",
        uniqueAddonName = "ShifterBoxExample",
        uniqueShifterBoxName = "ExampleFullWidth",
        shifterBoxCustomSettings = {
            showMoveAllButtons = true,
            dragDropEnabled = true,
            sortEnabled = true,
            sortBy = "value",
            leftList = {
                title = "Lefties",
                rowHeight = 24,
                rowTemplateName = "ShifterBoxEntryTemplate",
                emptyListText = GetString(LIBSHIFTERBOX_EMPTY),
                fontSize = 14,
            },
            rightList = {
                title = "Righties",
                rowHeight = 32,
                rowTemplateName = "ShifterBoxEntryTemplate",
                emptyListText = GetString(LIBSHIFTERBOX_EMPTY),
                fontSize = 18,
            }
        },
        leftListEntries = function() return {[1] = "Hello", [2] = "from", [3] = "the"} end,
        rightListEntries = {[4] = "other", [5] = "side"},
        width = "full",
        height = 50,
        disabled = function() return fullWidthShifterBoxDisabled end,
        reference = "SHIFTERBOXEXAMPLE_FULL",
    })

    OptionsTable:insert({
        type = "checkbox",
        name = "Dummy Entry",
        getFunc = function() return false end,
        setFunc = function() end,
        disabled = function() return fullWidthShifterBoxDisabled end,
    })

    OptionsTable:insert({
        type = "checkbox",
        name = "Disable above shifterBox?",
        getFunc = function() return fullWidthShifterBoxDisabled end,
        setFunc = function(value) fullWidthShifterBoxDisabled = value end,
    })

    OptionsTable:insert({
        type = "description",
        text = "And here in half-width:"
    })

    OptionsTable:insert({
        type = "shifterbox",
        uniqueAddonName = "ShifterBoxExample",
        uniqueShifterBoxName = "ExampleHalfWidth",
        shifterBoxCustomSettings = {
            showMoveAllButtons = true,
            dragDropEnabled = false,
            sortEnabled = false,
            sortBy = "value",
            leftList = {
                title = "Available",
            },
            rightList = {
                title = "Selected",
            }
        },
        leftListEntries = function() return {[1] = "The", [2] = "quick", [3] = "brown", [4] = "fox", [5] = "jumps", [6] = "over", [7] = "the", [8] = "lazy", [9] = "dog"} end,
        width = "half",
        height = 400,
        reference = "SHIFTERBOXEXAMPLE_HALF",
    })

    OptionsTable:insert({
        type = "description",
        text = "<-- This is a LibShifterBox half-width",
        width = "half",
    })

    -- ----------------------------------------------

    OptionSubTable:insert({
        type = "description",
        text = "And this is a sub-level, half-width shifterbox:"
    })

    OptionSubTable:insert({
        type = "description",
        text = "This is a sub-level LibShifterBox half-width -->",
        width = "half",
    })

    OptionSubTable:insert({
        type = "shifterbox",
        uniqueAddonName = "ShifterBoxExample",
        uniqueShifterBoxName = "ExampleNestedHalfWidth",
        shifterBoxCustomSettings = {
            showMoveAllButtons = true,
            dragDropEnabled = false,
            sortEnabled = false,
            sortBy = "value",
            leftList = {
                title = "Available",
            },
            rightList = {
                title = "Selected",
            }
        },
        leftListEntries = function() return {[1] = "The", [2] = "quick", [3] = "brown", [4] = "fox", [5] = "jumps", [6] = "over", [7] = "the", [8] = "lazy", [9] = "dog"} end,
        width = "half",
        height = 200,
        reference = "SHIFTERBOXEXAMPLE_NESTED_HALF",
    })

    OptionsTable:insert({
        type = "submenu",
        name = "Shifterbox in sub-level menu",
        controls = OptionSubTable,
    })

    LAM2:RegisterAddonPanel("ShifterBoxExampleAddonOptions", PanelData)
    LAM2:RegisterOptionControls("ShifterBoxExampleAddonOptions", OptionsTable)
end

local function initAddon(_, addOnName)
    if addOnName ~= _addonName then
        return
    end

    EVENT_MANAGER:UnregisterForEvent("ShifterBoxExampleAddonInit", EVENT_ADD_ON_LOADED)

    initShifterBoxExample()

    initLAMMenuWithShifterBoxExample()
end

EVENT_MANAGER:RegisterForEvent("ShifterBoxExampleAddonInit", EVENT_ADD_ON_LOADED, initAddon)
